File: mgmt2/rpc/jsonrpc/JsonRpc.cc
/**
  @section license License

  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

#include "JsonRpc.h"

#include <iostream>
#include <chrono>
#include <system_error>

#include "json/YAMLCodec.h"

namespace rpc
{
static constexpr auto logTag = "rpc";

namespace error = jsonrpc::error;

JsonRpc::Dispatcher::response_type
JsonRpc::Dispatcher::dispatch(jsonrpc::RpcRequestInfo const &request) const
{
  if (request.is_notification()) {
    std::error_code ec;
    invoke_notification_handler(request, ec);
    if (ec) {
      return {std::nullopt, ec};
    }
  } else {
    return invoke_handler(request);
  }

  return {};
}

jsonrpc::MethodHandler
JsonRpc::Dispatcher::find_handler(const std::string &methodName) const
{
  std::lock_guard<std::mutex> lock(_mutex);

  auto search = _methods.find(methodName);
  return search != std::end(_methods) ? search->second : nullptr;
}

jsonrpc::NotificationHandler
JsonRpc::Dispatcher::find_notification_handler(const std::string &notificationName) const
{
  std::lock_guard<std::mutex> lock(_mutex);

  auto search = _notifications.find(notificationName);
  return search != std::end(_notifications) ? search->second : nullptr;
}

JsonRpc::Dispatcher::response_type
JsonRpc::Dispatcher::invoke_handler(jsonrpc::RpcRequestInfo const &request) const
{
  auto handler = find_handler(request.method);
  if (!handler) {
    return {std::nullopt, error::RpcErrorCode::METHOD_NOT_FOUND};
  }
  jsonrpc::RpcResponseInfo response{request.id};

  try {
    auto const &rv = handler(*request.id, request.params);

    if (rv.isOK()) {
      response.callResult.result = rv.result();
    } else {
      // if we have some errors to log, then include it.
      response.callResult.errata = rv.errata();
    }
  } catch (std::exception const &e) {
    Debug(logTag, "Ups, something happened during the callback invocation: %s", e.what());
    return {std::nullopt, error::RpcErrorCode::INTERNAL_ERROR};
  }

  return {response, {}};
}

void
JsonRpc::Dispatcher::invoke_notification_handler(jsonrpc::RpcRequestInfo const &notification, std::error_code &ec) const
{
  auto handler = find_notification_handler(notification.method);
  if (!handler) {
    ec = error::RpcErrorCode::METHOD_NOT_FOUND;
    return;
  }
  try {
    handler(notification.params);
  } catch (std::exception const &e) {
    Debug(logTag, "Ups, something happened during the callback(notification) invocation: %s", e.what());
    // it's a notification so we do not care much.
  }
}

bool
JsonRpc::Dispatcher::remove_handler(std::string const &name)
{
  std::lock_guard<std::mutex> lock(_mutex);
  auto foundIt = std::find_if(std::begin(_methods), std::end(_methods), [&](auto const &p) { return p.first == name; });
  if (foundIt != std::end(_methods)) {
    _methods.erase(foundIt);
    return true;
  }

  return false;
}

bool
JsonRpc::Dispatcher::remove_notification_handler(std::string const &name)
{
  std::lock_guard<std::mutex> lock(_mutex);
  auto foundIt = std::find_if(std::begin(_notifications), std::end(_notifications), [&](auto const &p) { return p.first == name; });
  if (foundIt != std::end(_notifications)) {
    _notifications.erase(foundIt);
    return true;
  }

  return false;
}

// RPC

void
JsonRpc::register_internal_api()
{
  // register. TMP
  if (!_dispatcher.add_handler("show_registered_handlers",
                               // TODO: revisit this.
                               [this](std::string_view const &id, const YAML::Node &req) -> ts::Rv<YAML::Node> {
                                 return _dispatcher.show_registered_handlers(id, req);
                               })) {
    Warning("Something happened, we couldn't register the rpc show_registered_handlers handler");
  }
}

bool
JsonRpc::remove_handler(std::string const &name)
{
  return _dispatcher.remove_handler(name);
}

bool
JsonRpc::remove_notification_handler(std::string const &name)
{
  return _dispatcher.remove_notification_handler(name);
}

static inline jsonrpc::RpcResponseInfo
make_error_response(jsonrpc::RpcRequestInfo const &req, std::error_code const &ec)
{
  jsonrpc::RpcResponseInfo resp;

  // we may have been able to collect the id, if so, use it.
  if (req.id) {
    resp.id = req.id;
  }

  resp.rpcError = ec; // std::make_optional<jsonrpc::RpcError>(ec.value(), ec.message());

  return resp;
}

static inline jsonrpc::RpcResponseInfo
make_error_response(std::error_code const &ec)
{
  jsonrpc::RpcResponseInfo resp;

  resp.rpcError = ec;
  // std::make_optional<jsonrpc::RpcError>(ec.value(), ec.message());
  return resp;
}

std::optional<std::string>
JsonRpc::handle_call(std::string_view request)
{
  Debug(logTag, "Incoming request '%s'", request.data());

  std::error_code ec;
  try {
    // let's decode all the incoming messages into our own types.
    jsonrpc::RpcRequest const &msg = Decoder::decode(request, ec);

    // If any error happened within the request, they will be kept inside each
    // particular request, as they would need to be converted back in a proper error response.
    if (ec) {
      auto response = make_error_response(ec);
      return Encoder::encode(response);
    }

    jsonrpc::RpcResponse response{msg.is_batch()};
    for (auto const &[req, decode_error] : msg.get_messages()) {
      // As per jsonrpc specs we do care about invalid messages as long as they are well-formed,  our decode logic will make their
      // best to build up a request, if any errors were detected during decoding, we will save the error and make it part of the
      // RPCRequest elements for further use.
      if (!decode_error) {
        // request seems ok and ready to be dispatched. The dispatcher will tell us if the method exist and if so, it will dispatch
        // the call and gives us back the response.
        auto &&[encodedResponse, ec] = _dispatcher.dispatch(req);

        // On any error, ec will have a value
        if (!ec) {
          // we only get valid responses if it was a method request, not
          // for notifications.
          if (encodedResponse) {
            response.add_message(*encodedResponse);
          }
        } else {
          // get an error response, we may have the id, so let's try to use it.
          response.add_message(make_error_response(req, ec));
        }

      } else {
        // If the request was marked as error(decode error), we still need to send the error back, so we save it.
        response.add_message(make_error_response(req, decode_error));
      }
    }

    // We will not have a response for notification(s); This could be a batch of notifications only.
    return response.is_notification() ? std::nullopt : std::make_optional<std::string>(Encoder::encode(response));

  } catch (std::exception const &ex) {
    ec = jsonrpc::error::RpcErrorCode::INTERNAL_ERROR;
  }

  return {Encoder::encode(make_error_response(ec))};
}

ts::Rv<YAML::Node>
JsonRpc::Dispatcher::show_registered_handlers(std::string_view const &, const YAML::Node &)
{
  ts::Rv<YAML::Node> resp;
  // use the same lock?
  std::lock_guard<std::mutex> lock(_mutex);
  for (auto const &m : _methods) {
    std::string sm = m.first;
    resp.result()["methods"].push_back(sm);
  }

  for (auto const &m : _notifications) {
    std::string sm = m.first;
    resp.result()["notifications"].push_back(sm);
  }

  return resp;
}
} // namespace rpc

File: tests/gold_tests/autest-site/jsonrpc.py
'''
JSONRPC Request convenient class helper.
'''
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF typing.ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import typing
from collections import OrderedDict
import json
import uuid


class BaseRequestType(type):
    '''
    Base class for both, request and notifications
    '''
    def __getattr__(cls: typing.Callable, name: str) -> typing.Callable:
        def attr_handler(*args: typing.Any, **kwargs: typing.Any) -> "Request":
            return cls(name, *args, **kwargs)
        return attr_handler


class Notification(dict, metaclass=BaseRequestType):
    '''
    Convenient class to create JSONRPC objects(Notifications) and strings without the need to specify redundant information.

    Examples:

        Notification.foo_bar() = > {"jsonrpc": "2.0", "method": "foo_bar"}

        Notification.foo_bar({"hello"="world"}) =>
            {
                    "jsonrpc": "2.0",
                    "method": "foo_bar",
                    "params": {
                        "hello": "world"
                    }
                }

        Notification.foo_bar(var={"hello":"world"}) =>
            {
                "jsonrpc": "2.0",
                "method": "foo_bar",
                "params": {
                    "var": {
                        "hello": "world"
                    }
                }
            }
        Notification.foo_bar(fqdn=["yahoo.com", "trafficserver.org"]) =>
            {
                "jsonrpc": "2.0",
                "method": "foo_bar",
                "params": {
                    "fqdn": ["yahoo.com", "trafficserver.org"]
                }
            }

    '''

    def __init__(self, method: str, *args: typing.Any, **kwargs: typing.Any):
        super(Notification, self).__init__(jsonrpc='2.0', method=method)
        if args and kwargs:
            plist = list(args)
            plist.append(kwargs)
            self.update(params=plist)
        elif args:
            if isinstance(args, tuple):
                # fix this. this is to avoid having params=[[values] which seems invalid. Double check this as [[],[]] would be ok.
                self.update(params=args[0])
            else:
                self.update(params=list(args))
        elif kwargs:
            self.update(params=kwargs)

    # Allow using the dict item as an attribute.
    def __getattr__(self, name):
        if name in self:
            return self[name]
        else:
            raise AttributeError("No such attribute: " + name)

    def is_notification(self):
        return True

    def __str__(self) -> str:
        return json.dumps(self)


class Request(Notification):
    '''
    Convenient class to create JSONRPC objects and strings without the need to specify redundant information like
    version or the id. All this will be generated automatically.

    Examples:

        Request.foo_bar() = > {"id": "e9ad55fe-d5a6-11eb-a2fd-fa163e6d2ec5", "jsonrpc": "2.0", "method": "foo_bar"}

        Request.foo_bar({"hello"="world"}) =>
            {
                    "id": "850d2998-d5a7-11eb-bebc-fa163e6d2ec5",
                    "jsonrpc": "2.0",
                    "method": "foo_bar",
                    "params": {
                        "hello": "world"
                    }
                }

        Request.foo_bar(var={"hello":"world"}) =>
            {
                "id": "850d2e84-d5a7-11eb-bebc-fa163e6d2ec5",
                "jsonrpc": "2.0",
                "method": "foo_bar",
                "params": {
                    "var": {
                        "hello": "world"
                    }
                }
            }
        Request.foo_bar(fqdn=["yahoo.com", "trafficserver.org"]) =>
            {
                "id": "850d32a8-d5a7-11eb-bebc-fa163e6d2ec5",
                "jsonrpc": "2.0",
                "method": "foo_bar",
                "params": {
                    "fqdn": ["yahoo.com", "trafficserver.org"]
                }
            }

    Note: Use full namespace to avoid name collision => jsonrpc.Request, jsonrpc.Response, etc.
    '''

    def __init__(self, method: str, *args: typing.Any, **kwargs: typing.Any):
        if 'id' in kwargs:
            self.update(id=kwargs.pop('id'))  # avoid duplicated
        else:
            self.update(id=str(uuid.uuid1()))

        super(Request, self).__init__(method, *args, **kwargs)

    def is_notification(self):
        return False


class BatchRequest(list):
    def __init__(self, *args: typing.Union[Request, Notification]):
        for r in args:
            self.append(r)

    def add_request(self, req: typing.Union[Request, Notification]):
        self.append(req)

    def __str__(self) -> str:
        return json.dumps(self)


class Response(dict):
    '''
    Convenient class to help handling jsonrpc responses. This can be source from text directly or by an already parsed test into
    a json object.
    '''

    def __init__(self, *arg, **kwargs):
        if 'text' in kwargs:
            self.__dict__ = json.loads(kwargs['text'])
        elif 'json' in kwargs:
            self.__dict__ = kwargs['json']

    def is_error(self) -> bool:
        '''
        Check whether the error field is present in the response. ie:
            {
                "jsonrpc":"2.0",
                "error":{...},
                "id":"284e0b86-d03a-11eb-9206-fa163e6d2ec5"
            }
        '''
        if 'error' in self.__dict__:
            return True
        return False

    def is_only_success(self) -> bool:
        '''
        Some responses may only have set the result as success. This functions checks that the value in the response field only
        contains the 'success' string, ie:

            {
                "jsonrpc": "2.0",
                "result": "success",
                "id": "8504569c-d5a7-11eb-bebc-fa163e6d2ec5"
            }
        '''
        if self.is_ok() and self.result == 'success':
            return True

        return False

    def is_ok(self) -> bool:
        '''
        No error present in the response, the result field was properly set.
        '''
        return 'result' in self.__dict__

    def error_as_str(self):
        '''
        Build up the error string.
        {
            "jsonrpc":"2.0",
            "error":{
                "code":9,
                "message":"Error during execution",
                "data":[
                    {
                        "code":10001,
                        "message":"No values provided"
                    }
                ]
            },
            "id":"284e0b86-d03a-11eb-9206-fa163e6d2ec5"
            }
        '''
        if self.is_ok():
            return "no error"

        errStr = ""
        errStr += f"\n[code: {self.error['code']}, message: \"{self.error['message']}\"]"
        if 'data' in self.error:
            errStr += "\n\tAdditional Information:"
            for err in self.error['data']:
                errStr += f"\n\t - code: {err['code']}, message: \"{err['message']}\"."
        return errStr

    def __str__(self) -> str:
        return json.dumps(self.__dict__)

    def is_execution_error(self):
        '''
        Checks if the provided error is an Execution error(9).
        '''
        if self.is_ok():
            return False

        return self.error['code'] == 9

    def contains_nested_error(self, code=None, msg=None):
        if self.is_execution_error():
            for err in self.error['data']:
                if code and msg:
                    return err['code'] == code and err['message'] == msg
                elif code and err['code'] == code:
                    return True
                elif msg and err['message'] == msg:
                    return True
                else:
                    return False
        return False


def make_response(text):
    if text == '':
        return None

    s = json.loads(text)
    if isinstance(s, dict):
        return Response(json=s)
    elif isinstance(s, list):
        batch = []
        for r in s:
            batch.append(Response(json=r))
        return batch
